binder/src/main/java/io/grpc/binder/internal/PendingAuthListener.java

```diff
@@ -0,0 +1,156 @@
+package io.grpc.binder.internal;
+
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import io.grpc.Metadata;
+import io.grpc.ServerCall;
+import io.grpc.ServerCallHandler;
+import io.grpc.Status;
+import io.grpc.internal.ObjectPool;
+
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicReference;
+
+import javax.annotation.Nullable;
+
+/**
+ * A {@link ServerCall.Listener} that can be returned by a {@link io.grpc.ServerInterceptor} to
+ * asynchronously advance the gRPC pending resolving a possibly asynchronous security policy check.
+ */
+final class PendingAuthListener<ReqT, RespT> extends ServerCall.Listener<ReqT> {
+
+  private final ConcurrentLinkedQueue<ListenerConsumer<ReqT>> pendingSteps =
+      new ConcurrentLinkedQueue<>();
+  private final AtomicReference<ServerCall.Listener<ReqT>> delegateRef =
+      new AtomicReference<>(null);
+
+  /**
+   * @param authStatusFuture a ListenableFuture holding the result status of the authorization
+   *                         policy from a {@link io.grpc.binder.SecurityPolicy} or a
+   *                         {@link io.grpc.binder.AsyncSecurityPolicy}. The call only progresses
+   *                         if {@link Status#isOk()} is true.
+   * @param executorPool     a pool that can provide at least one Executor under which the result
+   *                         of {@code authStatusFuture} can be handled, progressing the gRPC
+   *                         stages.
+   * @param call             the 'call' parameter from {@link io.grpc.ServerInterceptor}
+   * @param headers          the 'headers' parameter from {@link io.grpc.ServerInterceptor}
+   * @param next             the 'next' parameter from {@link io.grpc.ServerInterceptor}
+   */
+  static <ReqT, RespT> PendingAuthListener<ReqT, RespT> create(
+      ListenableFuture<Status> authStatusFuture,
+      ObjectPool<? extends Executor> executorPool,
+      ServerCall<ReqT, RespT> call,
+      Metadata headers,
+      ServerCallHandler<ReqT, RespT> next) {
+    PendingAuthListener<ReqT, RespT> listener = new PendingAuthListener<>();
+    Executor executor = executorPool.getObject();
+    Futures.addCallback(
+        authStatusFuture,
+        new FutureCallback<Status>() {
+          @Override
+          public void onSuccess(Status authStatus) {
+            try {
+              if (!authStatus.isOk()) {
+                call.close(authStatus, new Metadata());
+                return;
+              }
+
+              listener.startCall(call, headers, next);
+            } finally {
+              executorPool.returnObject(executor);
```

Comment:
I think I'm slightly uncomfortable with the way we're managing the executorpool in this class.

Instead, I'd suggest we should handle that elsewhere and have this class only use the executor.

I'm thinking 'elsewhere' should be either inside the interceptor or BinderServer.

Reply:
It's never actually been clear to me what ObjectPool<> is doing for us. Like, the Executor interface is already safe for multiple threads to use concurrently. Mark do you happen to know why we need to check-out Executors from the pool and then ensure they are returned? Users configure their Executors on the ChannelBuilder or ServerBuilder and implicitly promise not to shut them down for the lifetime of those objects -- so can't we just use them freely?

Reply:
I think it's just about knowing when the executor can be shut down, which is only relevant when gRPC itself provides the executors.

Reply:
Updated.